colossalai/zero/gemini/gemini_optimizer.py

```diff
@@ -301,25 +308,365 @@ def get_range_pair(local_chunk: Chunk, local_param: Parameter):
+        dist.barrier(process_group)
+        range_info = dict()
+        if rank == master_rank:
+
+            # Record if master owns this param.
+            if state_shard_range[-1] >= 0:
+                range_info[master_rank] = (state_shard_range[-2].item(), state_shard_range[-1].item())
+
+            container_tensor = torch.zeros_like(state_shard_range, dtype=torch.int, requires_grad=False).cuda()
+            for src in range(0, dist.get_world_size(process_group)):
+                if src == master_rank:
+                    continue
+                dist.recv(container_tensor, src, process_group)
+                # Only keeps valid position messages.
+                if container_tensor[-1] >= 0:
+                    range_info[container_tensor[0].item()] = (container_tensor[-2].item(), container_tensor[-1].item())
+            assert len(list(range_info.keys())) > 0
+
+        else:
+            dist.send(state_shard_range, master_rank, process_group)
+        dist.barrier(process_group)
+
+        # Master get prepared for state collecting.
+        if rank == master_rank:
+            for state_name in state_names:
+                if state_name == 'step':
+                    # To keep aligned with pytorch, state 'step' is stored as a pytorch tensor with type float32.
+                    collected_states[state_name] = torch.tensor(0.0, dtype=torch.float32, requires_grad=False).cpu()
+                else:
+                    collected_states[state_name] = torch.zeros(param.numel(), dtype=torch.float32,
+                                                               requires_grad=False).cpu()
+
+        # Master starts to receive state tensors from other ranks
+        # after all the position messages of state shards has been collected.
+        compacted_states = self.pack_optimizer_states_to_tensor(param_id)
+        if rank == master_rank:
+            for src, shard_range in range_info.items():
+
+                # If src is master, directly load from compacted_states tensor.
+                if src == master_rank:
+                    self.load_from_compacted_states(compacted_states, collected_states, shard_range)
+                    continue
+
+                # If src is not master, receive states from other ranks.
+                shard_size = shard_range[1] - shard_range[0]
+                num_states = len(state_names)
+                container_size = 1 + (num_states - 1) * shard_size if 'step' in state_names \
+                                        else num_states * shard_size
+                container_tensor = torch.zeros(container_size, dtype=torch.float32, requires_grad=False).cuda()
+                dist.recv(container_tensor, src, process_group)
+                self.load_from_compacted_states(container_tensor, collected_states, shard_range)
+                del container_tensor
+        else:
+            if own_param:
+                dist.send(compacted_states, master_rank, process_group)
```

Comment:
Same problem. Multiple send/recv is inefficient

Reply:
Same response as the last conversation.